site/tests/app/libraries/routers/WebRouterTester.php

```diff
@@ -136,4 +138,34 @@ public function randomUrlProvider() {
             ["/s19/sample/random/invalid/endpoint"]
         ];
     }
+
+    /**
+     * @param string $url a url to an nonexistent API endpoint
+     * @dataProvider randomUrlProvider
+     */
+    public function testApiNotFound($url) {
+        $core = $this->createMockCore(['semester' => 's19', 'course' => 'sample']);
+        $request = Request::create(
+            "/api" . $url
+        );
+        $this->expectException(ResourceNotFoundException::class);
+        $router = new WebRouter($request, $core, true, true);
+        $this->assertEquals([
+            'status' => "fail",
+            'message' => "Endpoint not found."
+        ], $router->run()->render());
+    }
+
+    public function testApiWrongMethod() {
+        $core = $this->createMockCore(['semester' => 's19', 'course' => 'sample']);
+        $request = Request::create(
+            "/api/token"
+        );
+        $this->expectException(MethodNotAllowedException::class);
```

Comment:
What method here throws this exception? I'm not sure I follow the logic for how this gets raised?

Reply:
It is thrown by the matcher of symfony (located in the constructor) when the resource was found but the request method is not allowed.

Reply:
But you catch it internally and set a response object instead right?

Reply:
Yes

Reply:
Well it looks that it will make more sense if the test method calls the static method. I'll change it now.